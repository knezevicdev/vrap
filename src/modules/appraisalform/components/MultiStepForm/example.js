import React from 'react';
import useForm from '@app/components/Form/useForm';
import FirstNameInput from '@app/components/Form/Inputs/FirstNameInput';
import LastNameInput from '@app/components/Form/Inputs/LastNameInput';
import EmailInput from '@app/components/Form/Inputs/EmailInput';
import MultiStepForm from './';
import PropTypes from 'prop-types';

export const FirstInfoForm = ({ fields }) => (
  <>
    <FirstNameInput field={fields.first} />
    <LastNameInput field={fields.last} />
    <EmailInput field={fields.email} />
  </>
);

export const SecondInfoForm = ({ fields }) => (
  <>
    <FirstNameInput field={fields.first} />
    <LastNameInput field={fields.last} />
    <EmailInput field={fields.email} />
  </>
);

export const ThirdInfoForm = ({ fields }) => (
  <>
    <FirstNameInput field={fields.first} />
    <LastNameInput field={fields.last} />
    <EmailInput field={fields.email} />
  </>
);

FirstInfoForm.propTypes = {
  fields: PropTypes.object
};

SecondInfoForm.propTypes = {
  fields: PropTypes.object
};

ThirdInfoForm.propTypes = {
  fields: PropTypes.object
};

const FormMain = () => {
  const sendToRedux = () => {
    // console.log('reduxed');
  };

  const defaultFieldValues = {
    defaultValues: {
      first: '',
      last: '',
      email: ''
    },
    customEvents: {
      onBlur: sendToRedux
    }
  };

  const firstForm = useForm(defaultFieldValues);
  const secondForm = useForm(defaultFieldValues);
  const thirdForm = useForm(defaultFieldValues);

  const sections = [
    {
      component: FirstInfoForm,
      form: firstForm,
      subTitle: 'Form 1',
      title: 'Section 1'
    },
    {
      component: SecondInfoForm,
      form: secondForm,
      subTitle: 'Form 2',
      title: 'Section 2'
    },
    {
      component: ThirdInfoForm,
      form: thirdForm,
      subTitle: 'Form 3',
      title: 'Section 3'
    }
  ];

  const onSubmit = () => {
    // get field values for each "form"
    //     const firstFormFields = firstForm.fields;
    //     const secondFormFields = secondForm.fields;
    //     const thirdFormFields = thirdForm.fields;

    alert("Here is where we'd submit to api and add to redux if necessary");
  };

  return (
    <MultiStepForm
      formTitle="Multi-step form example"
      sections={sections}
      onDone={onSubmit}
      active={0}
    />
  );
};

export default FormMain;
