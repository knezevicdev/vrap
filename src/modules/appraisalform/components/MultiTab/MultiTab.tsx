import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';

const MultiTab = ({ tabSections, active = 0, className, theme }) => {
  const MultiTabTheme = theme;
  const [activeTab, setActiveTab] = useState(active);
  const tabWidth = 100 / tabSections.length;

  useEffect(
    () => {
      setActiveTab(active);
    },
    [active]
  );

  const handleTabClick = tabToOpen => {
    setActiveTab(tabToOpen);
  };

  const tabs = tabSections.map((tab, idx) => {
    const isActive = idx === activeTab;

    return (
      <MultiTabTheme.tabButton
        isActive={isActive}
        key={idx}
        tabWidth={tabWidth}
        onClick={() => handleTabClick(idx)}
        data-qa={tab.title}
      >
        {tab.title}
      </MultiTabTheme.tabButton>
    );
  });

  const tabComponents = tabSections.map((tab, idx) => {
    const CurrentTab = tab.component;
    const isActive = idx === activeTab;

    return (
      <MultiTabTheme.tabContentSection isActive={isActive} key={idx}>
        <CurrentTab {...tab.props} handleTabClick={handleTabClick} />
      </MultiTabTheme.tabContentSection>
    );
  });

  return (
    <MultiTabTheme className={className}>
      <MultiTabTheme.tabs>{tabs}</MultiTabTheme.tabs>
      <MultiTabTheme.body>{tabComponents}</MultiTabTheme.body>
    </MultiTabTheme>
  );
};

MultiTab.propTypes = {
  tabSections: PropTypes.array.isRequired,
  active: PropTypes.number,
  className: PropTypes.string,
  theme: PropTypes.object
};

export default MultiTab;
