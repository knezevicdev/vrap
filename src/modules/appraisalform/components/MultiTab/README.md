### MultiTab Component

The purpose of this component is to create an easy to use component that renders a tabbed component

### Required Props

1. tabSections - Collection; this is an array of each tab you'll be passing to the component.
   Each tabSection should contain a component and title.
   Component is the actual component you'll be rendered for that step.
   Prop is if your tab requires any custom props; some cases might include something else on the page changing tab data
   Title is to be a title for the whole section
   ex: {
   component: FirstInfoTab,
   props: { text: 'custom text' },
   title: 'Section 1'
   }
2. active - Integer that dictates which tab is shown first; defaults to 0

### Example usage of MultiTab

<MultiTab
  tabSections={tabSections}
  active={0}
/>

### Unit Tests

As of now there is a snapshot test written, so far I don't see a need for any more but as this component extends we should also extend the tests

### To view example.js - after adding it to src/routes.js

url: /multiTabComponent
