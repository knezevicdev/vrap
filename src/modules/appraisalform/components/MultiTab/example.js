import React from 'react';
import PropTypes from 'prop-types';
import MultiTab from './';

export const FirstInfoTab = () => (
  <>
    <span>tab 1</span>
  </>
);

export const SecondInfoTab = ({ props }) => (
  <>
    <span>tab 2</span>
    {props.text}
  </>
);

const TabMain = () => {
  const tabSections = [
    {
      component: FirstInfoTab,
      title: 'tab 1'
    },
    {
      component: SecondInfoTab,
      props: { text: 'passed in custom prop' },
      title: 'tab 2'
    }
  ];

  return <MultiTab tabSections={tabSections} active={0} />;
};

SecondInfoTab.propTypes = {
  props: PropTypes.object,
  text: PropTypes.string
};

export default TabMain;
