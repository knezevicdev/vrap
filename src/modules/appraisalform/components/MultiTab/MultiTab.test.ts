import MultiTab from './';
import { FirstInfoTab, SecondInfoTab } from './example';

describe('<MultiTab />', () => {
  let props;
  let wrapper;

  beforeEach(() => {
    const tabSections = [
      {
        component: FirstInfoTab,
        title: 'Section 1'
      },
      {
        component: SecondInfoTab,
        title: 'Section 2'
      }
    ];

    props = {
      tabSections: tabSections,
      active: 0
    };
  });

  describe('rendering', () => {
    it('should match the snapshot', () => {
      wrapper = shallow(<MultiTab {...props} />);
      expect(wrapper).toMatchSnapshot();
    });
  });
});
