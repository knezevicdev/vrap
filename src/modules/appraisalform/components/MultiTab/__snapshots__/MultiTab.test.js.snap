// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MultiTab /> rendering should match the snapshot 1`] = `
<styled.div>
  <styled.div>
    <styled.div
      isActive={true}
      key="0"
      onClick={[Function]}
      tabWidth={50}
    >
      Section 1
    </styled.div>
    <styled.div
      isActive={false}
      key="1"
      onClick={[Function]}
      tabWidth={50}
    >
      Section 2
    </styled.div>
  </styled.div>
  <styled.div>
    <styled.div
      isActive={true}
      key="0"
    >
      <FirstInfoTab />
    </styled.div>
    <styled.div
      isActive={false}
      key="1"
    >
      <SecondInfoTab />
    </styled.div>
  </styled.div>
</styled.div>
`;
